src/Features/CSharp/Portable/UseIndexOrRangeOperator/CSharpUseIndexOperatorDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,224 @@
+            OperationAnalysisContext context, InfoCache infoCache)
+        {
+            var cancellationToken = context.CancellationToken;
+            var propertyReference = (IPropertyReferenceOperation)context.Operation;
+
+            // Only analyze indexer calls.
+            if (!propertyReference.Property.IsIndexer)
+            {
+                return;
+            }
+
+            if (propertyReference.Arguments.Length != 1)
+            {
+                return;
+            }
+
+            AnalyzeInvokedMember(
+                context, infoCache,
+                propertyReference.Instance,
+                propertyReference.Property.GetMethod,
+                propertyReference.Arguments[0].Value,
+                lengthLikePropertyOpt: null,
+                cancellationToken);
+        }
+
+        private void AnalyzeArrayElementReference(
+            OperationAnalysisContext context, InfoCache infoCache, IPropertySymbol arrayLengthProperty)
+        {
+            var cancellationToken = context.CancellationToken;
+            var arrayElementReference = (IArrayElementReferenceOperation)context.Operation;
+
+            // Has to be a single-dimensional element access.
+            if (arrayElementReference.Indices.Length != 1)
+            {
+                return;
+            }
+
+            AnalyzeInvokedMember(
+                context, infoCache,
+                arrayElementReference.ArrayReference,
+                targetMethodOpt: null,
+                arrayElementReference.Indices[0],
+                lengthLikePropertyOpt: arrayLengthProperty,
+                cancellationToken);
+        }
+
+        private void AnalyzeInvokedMember(
+            OperationAnalysisContext context, InfoCache infoCache, 
+            IOperation instance, IMethodSymbol targetMethodOpt, IOperation argumentValue,
+            IPropertySymbol lengthLikePropertyOpt, CancellationToken cancellationToken)
+        {
+            // look for `s[s.Length - value]` or `s.Get(s.Length- value)`.
+
+            // Needs to have the one arg for `s.Length - value`, and that arg needs to be
+            // a subtraction.
+            if (instance is null ||
+                !IsSubtraction(argumentValue, out var subtraction))
+            {
+                return;
+            }
+
+            if (!(subtraction.Syntax is BinaryExpressionSyntax binaryExpression))
+            {
+                return;
+            }
+
+            // Only supported on C# 8 and above.
+            var syntaxTree = binaryExpression.SyntaxTree;
+            var parseOptions = (CSharpParseOptions)syntaxTree.Options;
+            if (parseOptions.LanguageVersion < LanguageVersion.CSharp8)
+            {
+                return;
+            }
+
+            // Don't bother analyzing if the user doesn't like using Index/Range operators.
+            var optionSet = context.Options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
```

Comment:
Do analyzers have to process async APIs this way? Surprised a bit we didn't have an async hook to use here.

Reply:
They do.  This should hopefully be addressed by all the work going on to make EditorConfig info available int he compiler though! :)  Tagging @agocke  on that.